arch/riscv/core/userspace.S

```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright (c) 2020 RISE Research Institutes of Sweden <www.ri.se>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <arch/cpu.h>
+#include <toolchain.h>
+#include <linker/sections.h>
+#include <offsets_short.h>
+#include <arch/cpu.h>
+#include <sys/util.h>
+#include <kernel.h>
+
+#ifdef CONFIG_USERSPACE
+
+GTEXT(arch_user_string_nlen)
+GTEXT(rv_user_mode_enter)
+
+SECTION_FUNC(TEXT, rv_user_mode_enter)
+	li t1, MSTATUS_MPP_M
+	csrc mstatus, t1
+	li t1, MSTATUS_MPIE_EN
+	csrs mstatus, t1
+	csrw mepc, ra
+	mret
+
+/* size_t arch_user_string_nlen(const char *s, size_t maxsize, int *err_arg) */
+SECTION_FUNC(TEXT, arch_user_string_nlen)
```

Comment:
Can you give extensive detail on the approach you are taking here? The string passed in from user mode is completely untrusted, all other architectures have set up a true HW-based exception mechanism to safely handle faults. Here we have something that does not do that, are you convinced that there are no possible inputs which could get the kernel to crash?

Reply:
Thanks for the feedback, we rewrote the implementations for both `arch_user_string_nlen` and`arch_is_user_context` to use the exception handler macros in `include/exc_handle.h` as the both Arm and Arc do. The new code comments should clarify how it works.